File: gtracr/lib/gpu/igrf_table.cpp
#include "igrf_table.hpp"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <random>
#include <stdexcept>

#include "igrf.hpp"
#include "constants.hpp"

// ---------------------------------------------------------------------------
// generate_igrf_table
// ---------------------------------------------------------------------------
std::vector<float> generate_igrf_table(IGRF& igrf, TableParams& params) {
    const int Nr     = IGRF_NR;
    const int Ntheta = IGRF_NTHETA;
    const int Nphi   = IGRF_NPHI;

    // Radial range: 1 RE to 10 RE (in meters)
    const double r_min = constants::RE;           // 6,371,200 m
    const double r_max = 10.0 * constants::RE;    // 63,712,000 m

    params.r_min      = static_cast<float>(r_min);
    params.r_max      = static_cast<float>(r_max);
    params.log_r_min  = static_cast<float>(std::log(r_min));
    params.log_r_max  = static_cast<float>(std::log(r_max));
    params.Nr         = Nr;
    params.Ntheta     = Ntheta;
    params.Nphi       = Nphi;

    std::vector<float> table(IGRF_TABLE_SIZE, 0.0f);

    const int slice = Nr * Ntheta * Nphi;  // size of one component block

    for (int ir = 0; ir < Nr; ++ir) {
        // Log-spaced r: r[i] = r_min * (r_max/r_min)^(i/(Nr-1))
        double t = static_cast<double>(ir) / (Nr - 1);
        double r = r_min * std::pow(r_max / r_min, t);

        for (int itheta = 0; itheta < Ntheta; ++itheta) {
            double theta = itheta * constants::pi / (Ntheta - 1);

            for (int iphi = 0; iphi < Nphi; ++iphi) {
                double phi = iphi * 2.0 * constants::pi / Nphi;

                std::array<double, 3> bfield = igrf.values(r, theta, phi);

                int base = ir * Ntheta * Nphi + itheta * Nphi + iphi;
                table[0 * slice + base] = static_cast<float>(bfield[0]);  // Br
                table[1 * slice + base] = static_cast<float>(bfield[1]);  // Btheta
                table[2 * slice + base] = static_cast<float>(bfield[2]);  // Bphi
            }
        }
    }

    return table;
}

// ---------------------------------------------------------------------------
// table_lookup  (CPU version — trilinear interpolation)
// ---------------------------------------------------------------------------
std::array<float, 3> table_lookup(const float* table, const TableParams& params,
                                   float r, float theta, float phi) {
    const int Nr     = params.Nr;
    const int Nt     = params.Ntheta;
    const int Np     = params.Nphi;
    const int slice  = Nr * Nt * Np;

    // Guard against NaN/Inf/negative inputs from RK45 intermediate stages.
    // std::isfinite returns false for NaN and Inf; r <= 0 would produce
    // NaN from std::log.  Return zero field to let the adaptive step-size
    // controller reject the step.
    if (!std::isfinite(r) || !std::isfinite(theta) || !std::isfinite(phi) || r <= 0.0f) {
        return {0.0f, 0.0f, 0.0f};
    }

    // ---- radial index (log-spaced) ----------------------------------------
    float log_r = std::log(r);
    float fr = (log_r - params.log_r_min) / (params.log_r_max - params.log_r_min)
               * (Nr - 1);
    // Clamp so ir0 is in [0, Nr-2], ensuring ir0+1 < Nr for trilinear interp.
    int ir0 = static_cast<int>(std::max(0.0f, std::min(static_cast<float>(Nr - 2), std::floor(fr))));
    float wr = std::max(0.0f, std::min(1.0f, fr - static_cast<float>(ir0)));

    // ---- theta index (linear) ---------------------------------------------
    float ft = theta * static_cast<float>(Nt - 1) / static_cast<float>(constants::pi);
    int it0 = static_cast<int>(std::max(0.0f, std::min(static_cast<float>(Nt - 2), std::floor(ft))));
    float wt = std::max(0.0f, std::min(1.0f, ft - static_cast<float>(it0)));

    // ---- phi index (linear, periodic) -------------------------------------
    // Normalise phi to [0, 2π)
    const float two_pi = static_cast<float>(2.0 * constants::pi);
    phi = phi - two_pi * std::floor(phi / two_pi);
    float fp = phi * static_cast<float>(Np) / two_pi;
    int ip0 = static_cast<int>(std::max(0.0f, std::min(static_cast<float>(Np - 1), std::floor(fp)))) % Np;
    int ip1 = (ip0 + 1) % Np;   // periodic wrap
    float wp = std::max(0.0f, std::min(1.0f, fp - std::floor(fp)));

    // ---- helper lambda to index into the flat array -----------------------
    auto idx = [&](int comp, int ir, int it, int ip) -> int {
        return comp * slice + ir * Nt * Np + it * Np + ip;
    };

    // ---- trilinear interpolation over (r, theta, phi) ---------------------
    std::array<float, 3> result;
    for (int c = 0; c < 3; ++c) {
        float v000 = table[idx(c, ir0,   it0,   ip0)];
        float v001 = table[idx(c, ir0,   it0,   ip1)];
        float v010 = table[idx(c, ir0,   it0+1, ip0)];
        float v011 = table[idx(c, ir0,   it0+1, ip1)];
        float v100 = table[idx(c, ir0+1, it0,   ip0)];
        float v101 = table[idx(c, ir0+1, it0,   ip1)];
        float v110 = table[idx(c, ir0+1, it0+1, ip0)];
        float v111 = table[idx(c, ir0+1, it0+1, ip1)];

        float along_phi_00 = (1.0f - wp) * v000 + wp * v001;
        float along_phi_01 = (1.0f - wp) * v010 + wp * v011;
        float along_phi_10 = (1.0f - wp) * v100 + wp * v101;
        float along_phi_11 = (1.0f - wp) * v110 + wp * v111;

        float along_theta_0 = (1.0f - wt) * along_phi_00 + wt * along_phi_01;
        float along_theta_1 = (1.0f - wt) * along_phi_10 + wt * along_phi_11;

        result[c] = (1.0f - wr) * along_theta_0 + wr * along_theta_1;
    }
    return result;
}

// ---------------------------------------------------------------------------
// validate_igrf_table
// ---------------------------------------------------------------------------
double validate_igrf_table(const float* table, const TableParams& params,
                            IGRF& igrf, int N) {
    std::mt19937 rng(42);
    std::uniform_real_distribution<double> rand_r(params.r_min, params.r_max);
    std::uniform_real_distribution<double> rand_theta(0.0, constants::pi);
    std::uniform_real_distribution<double> rand_phi(0.0, 2.0 * constants::pi);

    double max_rel_err = 0.0;
    const double eps = 1e-30;  // guard against near-zero denominator

    for (int i = 0; i < N; ++i) {
        double r     = rand_r(rng);
        double theta = rand_theta(rng);
        double phi   = rand_phi(rng);

        std::array<double, 3> direct = igrf.values(r, theta, phi);
        std::array<float, 3>  interp = table_lookup(table, params,
                                                     static_cast<float>(r),
                                                     static_cast<float>(theta),
                                                     static_cast<float>(phi));

        for (int c = 0; c < 3; ++c) {
            double ref = direct[c];
            double err = std::abs(static_cast<double>(interp[c]) - ref);
            double rel = err / (std::abs(ref) + eps);
            if (rel > max_rel_err) {
                max_rel_err = rel;
            }
        }
    }

    return max_rel_err;
}

File: gtracr/lib/igrf_table.py
"""
Python-level IGRF lookup table.

Mirrors the logic in gtracr/lib/gpu/igrf_table.cpp using numpy so that the
tabulated field can be used from Python (e.g. via pTrajectoryTracer) without
requiring GPU support.

Grid layout (matches the C++ constants in igrf_table.hpp):
  Nr=64    log-spaced radial points  [1 RE … 10 RE]
  Ntheta=128  linear colatitude       [0 … π]
  Nphi=256    linear longitude        [0 … 2π)
  Layout:  table[comp, ir, itheta, iphi]  — component-major, shape (3, Nr, Ntheta, Nphi)
"""

import os
from datetime import date as _date

import numpy as np

from gtracr.lib.constants import EARTH_RADIUS
from gtracr.utils import ymd_to_dec

# Grid dimensions — keep in sync with igrf_table.hpp
NR = 64
NTHETA = 128
NPHI = 256

_R_MIN = EARTH_RADIUS  # 1 RE  in metres
_R_MAX = 10.0 * EARTH_RADIUS  # 10 RE in metres

_DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "data")


class IGRFTable:
    """
    Precomputed IGRF lookup table with trilinear interpolation.

    Provides the same ``values(r, theta, phi) -> (Br, Btheta, Bphi)``
    interface as ``MagneticField`` / ``IGRF13`` so it can be used as a
    drop-in replacement in ``pTrajectoryTracer``.

    Parameters
    ----------
    igrf_obj : _libgtracr.IGRF, optional
        An already-initialised C++ IGRF object.  When *None* a new one is
        constructed from the bundled igrf13.json using today's date.
    verbose : bool
        Print progress while building the table (default False).
    """

    def __init__(self, igrf_obj=None, verbose=False):
        if igrf_obj is None:
            from gtracr.lib._libgtracr import IGRF

            datapath = os.path.abspath(_DATA_DIR)
            dec_date = float(ymd_to_dec(str(_date.today())))
            igrf_obj = IGRF(datapath, dec_date)

        self._r_min = _R_MIN
        self._r_max = _R_MAX
        self._log_r_min = np.log(_R_MIN)
        self._log_r_max = np.log(_R_MAX)

        self._table = self._build(igrf_obj, verbose)

    # ------------------------------------------------------------------
    # Table construction
    # ------------------------------------------------------------------

    def _build(self, igrf, verbose):
        """Evaluate igrf.values() on every grid point; return (3,Nr,Nt,Np) float32 array."""
        table = np.empty((3, NR, NTHETA, NPHI), dtype=np.float32)

        # Precompute grid vectors
        t_vec = np.arange(NR, dtype=np.float64) / (NR - 1)  # 0..1
        r_vec = _R_MIN * (_R_MAX / _R_MIN) ** t_vec  # log-spaced [m]
        th_vec = np.arange(NTHETA, dtype=np.float64) / (NTHETA - 1) * np.pi
        ph_vec = np.arange(NPHI, dtype=np.float64) / NPHI * 2.0 * np.pi

        total = NR * NTHETA
        done = 0
        for ir, r in enumerate(r_vec):
            for itheta, theta in enumerate(th_vec):
                for iphi, phi in enumerate(ph_vec):
                    b = igrf.values(r, theta, phi)
                    table[0, ir, itheta, iphi] = b[0]
                    table[1, ir, itheta, iphi] = b[1]
                    table[2, ir, itheta, iphi] = b[2]
                done += 1
                if verbose and done % (total // 10) == 0:
                    print(f"  igrf_table: {100 * done // total}% done", flush=True)

        return table

    # ------------------------------------------------------------------
    # Lookup (trilinear interpolation)
    # ------------------------------------------------------------------

    def values(self, r, theta, phi):
        """
        Interpolate B-field from the table.

        Parameters
        ----------
        r, theta, phi : float
            Position in geocentric spherical coordinates.
            r in metres, theta in radians [0,π], phi in radians (any range).

        Returns
        -------
        tuple of float : (Br, Btheta, Bphi)  in Tesla
        """
        # ---- radial index (log-spaced) --------------------------------
        log_r = np.log(max(r, self._r_min))
        fr = (log_r - self._log_r_min) / (self._log_r_max - self._log_r_min) * (NR - 1)
        fr = float(np.clip(fr, 0.0, NR - 1 - 1e-6))
        ir0 = int(fr)
        wr = fr - ir0

        # ---- theta index (linear) ------------------------------------
        ft = theta * (NTHETA - 1) / np.pi
        ft = float(np.clip(ft, 0.0, NTHETA - 1 - 1e-6))
        it0 = int(ft)
        wt = ft - it0

        # ---- phi index (linear, periodic) ----------------------------
        two_pi = 2.0 * np.pi
        phi_w = phi - two_pi * np.floor(phi / two_pi)  # wrap to [0, 2π)
        fp = phi_w * NPHI / two_pi
        fp = float(np.clip(fp, 0.0, NPHI - 1e-6))
        ip0 = int(fp) % NPHI
        ip1 = (ip0 + 1) % NPHI
        wp = fp - int(fp)

        # ---- trilinear interpolation ---------------------------------
        t = self._table
        result = []
        for c in range(3):
            v000 = float(t[c, ir0, it0, ip0])
            v001 = float(t[c, ir0, it0, ip1])
            v010 = float(t[c, ir0, it0 + 1, ip0])
            v011 = float(t[c, ir0, it0 + 1, ip1])
            v100 = float(t[c, ir0 + 1, it0, ip0])
            v101 = float(t[c, ir0 + 1, it0, ip1])
            v110 = float(t[c, ir0 + 1, it0 + 1, ip0])
            v111 = float(t[c, ir0 + 1, it0 + 1, ip1])

            along_phi_00 = (1.0 - wp) * v000 + wp * v001
            along_phi_01 = (1.0 - wp) * v010 + wp * v011
            along_phi_10 = (1.0 - wp) * v100 + wp * v101
            along_phi_11 = (1.0 - wp) * v110 + wp * v111

            along_theta_0 = (1.0 - wt) * along_phi_00 + wt * along_phi_01
            along_theta_1 = (1.0 - wt) * along_phi_10 + wt * along_phi_11

            result.append((1.0 - wr) * along_theta_0 + wr * along_theta_1)

        return result[0], result[1], result[2]

    # ------------------------------------------------------------------
    # Validation helper
    # ------------------------------------------------------------------

    def validate(self, igrf_obj, n=10000, rng_seed=42):
        """
        Compare table interpolation against direct ``igrf_obj.values()`` at
        *n* random points.

        Returns
        -------
        max_rel_err : float
            Maximum relative error across all components and test points.
            Target: < 0.001 (0.1 %).
        """
        rng = np.random.default_rng(rng_seed)
        r_pts = rng.uniform(self._r_min, self._r_max, n)
        th_pts = rng.uniform(0.0, np.pi, n)
        ph_pts = rng.uniform(0.0, 2.0 * np.pi, n)

        eps = 1e-30
        max_rel = 0.0
        for r, theta, phi in zip(r_pts, th_pts, ph_pts):
            direct = igrf_obj.values(r, theta, phi)
            interp = self.values(r, theta, phi)
            for d, v in zip(direct, interp):
                rel = abs(v - d) / (abs(d) + eps)
                if rel > max_rel:
                    max_rel = rel

        return max_rel
